include/crocoddyl/multibody/actions/constraint-fwddyn.hxx

```diff
@@ -0,0 +1,196 @@
+  const Eigen::VectorBlock<const Eigen::Ref<const VectorXs>, Eigen::Dynamic> v = x.tail(state_->get_nv());
+
+  // Computing the forward dynamics with the holonomic constraints defined by the contact model
+
+  actuation_->calc(d->multibody.actuation, x, u);
+  // TODO: Add mu
+  d->xout = pinocchio::constraintDynamics(pinocchio_, d->pinocchio, q, v, d->multibody.actuation->tau, contacts_,
+                                          d->multibody.contacts);  //, mu_contacts);
+  pinocchio::jacobianCenterOfMass(pinocchio_, d->pinocchio, false);
+  // Computing the cost value and residuals
+  costs_->calc(d->costs, x, u);
+  d->cost = d->costs->cost;
+}
+
+template <typename Scalar>
+void DifferentialActionModelConstraintFwdDynamicsTpl<Scalar>::calcDiff(
+    const boost::shared_ptr<DifferentialActionDataAbstract>& data, const Eigen::Ref<const VectorXs>& x,
+    const Eigen::Ref<const VectorXs>& u) {
+  if (static_cast<std::size_t>(x.size()) != state_->get_nx()) {
+    throw_pretty("Invalid argument: "
+                 << "x has wrong dimension (it should be " + std::to_string(state_->get_nx()) + ")");
+  }
+  if (static_cast<std::size_t>(u.size()) != nu_) {
+    throw_pretty("Invalid argument: "
+                 << "u has wrong dimension (it should be " + std::to_string(nu_) + ")");
+  }
+
+  const std::size_t& nv = state_->get_nv();
+  const Eigen::VectorBlock<const Eigen::Ref<const VectorXs>, Eigen::Dynamic> q = x.head(state_->get_nq());
+  const Eigen::VectorBlock<const Eigen::Ref<const VectorXs>, Eigen::Dynamic> v = x.tail(nv);
+
+  DifferentialActionDataConstraintFwdDynamicsTpl<Scalar>* d =
+      static_cast<DifferentialActionDataConstraintFwdDynamicsTpl<Scalar>*>(data.get());
+
+  // Computing the dynamics derivatives
+  actuation_->calcDiff(d->multibody.actuation, x, u);
+  pinocchio::computeConstraintDynamicsDerivatives(
+      pinocchio_, d->pinocchio, contacts_, d->multibody.contacts, d->Fx.leftCols(nv), d->Fx.rightCols(nv),
+      d->pinocchio.ddq_dtau, d->pinocchio.dlambda_dq, d->pinocchio.dlambda_dv, d->pinocchio.dlambda_dtau);
+
+  d->Fu.noalias() = d->pinocchio.ddq_dtau * d->multibody.actuation->dtau_du;
+  costs_->calcDiff(d->costs, x, u);
+}
+
+template <typename Scalar>
+boost::shared_ptr<DifferentialActionDataAbstractTpl<Scalar> >
+DifferentialActionModelConstraintFwdDynamicsTpl<Scalar>::createData() {
+  boost::shared_ptr<DifferentialActionDataAbstractTpl<Scalar> > data =
+      boost::make_shared<DifferentialActionDataConstraintFwdDynamics>(this);
+  DifferentialActionDataConstraintFwdDynamics* d =
+      static_cast<DifferentialActionDataConstraintFwdDynamics*>(data.get());
+  pinocchio::initConstraintDynamics(pinocchio_, d->pinocchio, contacts_);
+
+  return data;
+}
+
+template <typename Scalar>
+bool DifferentialActionModelConstraintFwdDynamicsTpl<Scalar>::checkData(
+    const boost::shared_ptr<DifferentialActionDataAbstract>& data) {
+  boost::shared_ptr<DifferentialActionDataConstraintFwdDynamics> d =
```

Comment:
Use `const &` to avoid create unnecessarily an owner, i.e.,
```suggestion
  const boost::shared_ptr<Data>& d =
```

(I am aware that there are a few places with this issues, it is in my TODO task to clean them.)